src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,381 @@
+        auto& ioContext = IoContext::current();
+        // Wait for the RPC to resolve and then process the result.
+        return js.wrapSimplePromise(ioContext.awaitIo(js, sendWorkerRpc(js, methodName, args),
+            [](jsg::Lock& js, capnp::Response<rpc::CallResults> result) -> jsg::Value {
+          return handleRpcResponse(js, kj::mv(result));
+        }));
+      }
+  ));
+}
+
+// The capability that lets us call remote methods over RPC.
+// The client capability is dropped after each callRequest().
+class JsRpcTargetImpl final : public rpc::JsRpcTarget::Server {
+public:
+  JsRpcTargetImpl(
+      kj::Own<kj::PromiseFulfiller<void>> callFulfiller,
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext callContext) override {
+    auto methodName = kj::heapString(callContext.getParams().getMethodName());
+    auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+    // Try to execute the requested method.
+    return ctx.run(
+        [this, methodName=kj::mv(methodName), serializedArgs=kj::mv(serializedArgs),
+          entrypointName=entrypointName, callContext] (Worker::Lock& lock) mutable ->
+            kj::Promise<void> {
+
+      jsg::Lock& js = lock;
+      if (!FeatureFlags::get(js).getJsRpc()) {
+        // JS RPC is not enabled on the server side, we cannot call any methods.
+        auto rpcDisabled =
+            "The receiving Worker does not allow its methods to be called over RPC."_kjc;
+        auto serialized = serializeV8(js, js.error(rpcDisabled));
+        auto builder = callContext.initResults(
+            capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+        builder.setV8Serialized(kj::mv(serialized));
+        return kj::READY_NOW;
+      }
+
+      // We will try to get the function, if we can't we'll throw an error to the client.
+      auto maybeFn = tryGetFn(lock, entrypointName, ctx, methodName);
+      KJ_IF_SOME(e, maybeFn.tryGet<kj::String>()) {
+        auto serialized = serializeV8(js, js.error(e));
+        auto builder = callContext.initResults(
+            capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+        builder.setV8Serialized(kj::mv(serialized));
+        return kj::READY_NOW;
+      }
+
+      auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+
+      // We have a function, so let's call it and serialize the result for RPC.
+      // We account for both the JS method throwing and our serialization attempt failing.
+      // If the function returns a promise we will wait for the promise to finish so we can
+      // serialize the result.
+      // TODO(now): Do we need to use awaitJs here? `call()` has to return a kj::Promise, so maybe
```

Comment:
This `awaitJs()` is appropriate, since the lambda passed to `run()` is required to return a KJ Promise. Note that we don't attempt to take any redundant locks after that promise resolves so there's no issue here.